exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/UpsertStatement.kt

```diff
@@ -61,3 +64,79 @@ open class UpsertStatement<Key : Any>(
         return super.prepared(transaction, sql)
     }
 }
+
+/**
+ * Common interface for building SQL statements that either insert a new row into a table,
+ * or update the existing row if insertion would violate a unique constraint.
+ */
+sealed interface UpsertBuilder {
+    /**
+     * Calls the specified function [body] with an [UpsertBuilder] as its receiver and an [UpdateStatement]
+     * as its argument, allowing values to be assigned to the UPDATE clause of an upsert statement.
+     *
+     * To specify manually that the insert value should be used when updating a column, for example within an expression
+     * or function, invoke `insertValue()` with the desired column as the function argument.
+     *
+     * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testUpsertWithManualUpdateUsingInsertValues
+     */
+    fun onUpdate(body: UpsertBuilder.(UpdateStatement) -> Unit) {
+        val arguments = UpdateStatement((this as InsertStatement<*>).table, null).apply {
+            body.invoke(this@UpsertBuilder, this)
+        }.firstDataSet
+        when (this) {
+            is UpsertStatement<*> -> updateValues.putAll(arguments)
+            is BatchUpsertStatement -> updateValues.putAll(arguments)
+        }
```

Comment:
Ideally, `updateValues` would be a property of the interface so this could be simplified, but doing so would expose the mutable map publicly since properties can't be internal or private.